### PR TITLE
Refactor filename generation and validation

### LINKED ISSUE #99: Generated filename is invalid if component has hashes
Some of the commands generate a new filename for the SBOM when used with the `--output` option pointing to a folder.

This filename currently fails to validate if `metadata.component.hashes` exists and is not empty.

That's because the filename generation algorithm in `output.py` doesn't take hashes into consideration at all.

On the other hand, the validator requires a hash to be present if the metadata component has at least one hash.

I'd be willing to fix this if you let me know which of the two sides of the problem you'd like to change: Either the hashes are added to the filename generator, or the validator accepts filenames without hashes even if the metadata component has one.

### PR DESCRIPTION
**Pull request by mmarseu**
This PR overhauls filename validation so it no longer fails to validate names generated by the tool itself.

In order to avoid similar situations in the future, generation and validation routines are moved into a dedicated module where they live side-by-side and can be changed together.

Unit tests have also been rewritten and moved to a new file. There are two categories of tests for this module:
- Tests which generate a name for an SBOM and validate it immediately. These are intended to find discrepancies between generation and validation.
- Tests which only validate names. These are intended to test the correctness of the validator when faced with names **not** generated by the tool itself.

This was a little more work as expected, especially because during coding I stumbled across a number of edge cases which aren't clearly defined and I wasn't sure how to approach them.  
As a result, I cannot fully rule out breaking changes introduced by this PR, in the sense that some filenames might no longer validate which used to be valid before.
During review, please pay close attention to the tests. They reflect all the special cases I considered and if there's something missing, please let me know.

Closes #99

### CODE REVIEW

**Review by CBeck-96 — APPROVED** [review]
Tried several strange combinations (including special symbols), as long as they were legitimate possibilities, they all passed.

**Comment by mmarseu** [comment]
> Tried several strange combinations (including special symbols), as long as they were legitimate possibilities, they all passed.

Your comment brought something to my mind which I'd like to investigate before merging. I've converted the PR to a draft.

There is the possibility for bugs and even security vulnerabilities with certain special characters (think path separators) in the `.metadata.component.name` and `.metadata.component.version`.

**Review by CBeck-96 — DISMISSED** [review]

**Review by italvi — CHANGES_REQUESTED** [review]
@mmarseu I tried this with the following minimal SBOM called `bom.json`:
```json
{
    "$schema": "http://cyclonedx.org/schema/bom-1.4.schema.json",
    "bomFormat": "CycloneDX",
    "specVersion": "1.4",
    "serialNumber": "urn:uuid:78934cae-d936-4673-aabb-482aacae2c87",
    "version": 1,
    "metadata": {
        "timestamp": "2023-07-26T15:04:48.117371+00:00",
        "authors": [
            {
                "name": "automated"
            }
        ],
        "component": {
            "type": "firmware",
            "bom-ref": "bom_ref_test",
            "author": "Festo",
            "name": "test",
            "version": "1.5.30",
            "copyright": "Copyright by Festo SE",
            "hashes": [
                {
                    "alg": "MD5",
                    "content": "ec7781220ec7781220ec778122012345"
                }
            ]
        }
    },
    "components": [],
    "compositions": [
        {
            "aggregate": "incomplete",
            "assemblies": [
                "bom_ref_test"
            ]
        }
    ],
    "dependencies": [
        {
            "ref": "bom_ref_test",
            "dependsOn": []
        }
    ]
}
```

using the command `python -m cdxev amend bom.json --output=.` resulted in the following output: `Writing output to: test_1.5.30_20240302T110102.cdx.json`. Why is the `hashes[0].content` not in the filename?

File: cdxev/auxiliary/filename_gen.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+def _sanitize(s: str) -> str:
+    """
+    Converts a string to a representation safe for filenames.
+
+    The following transformations are done:
+    * Normalize Unicode
+    * Filter out any characters which aren't alphanumeric, spaces, dashes, periods or underscores.
+    * Strip any leading or trailing non-alphanumeric characters.
+
+    :param str value: The original string.
```
[the remark this answers is not in the document]
Comment (mmarseu) [reply cdxev/auxiliary/filename_gen.py]: You're clearly reading this very closely, I'm impressed 😆 
Fixed in 9158af3

File: cdxev/auxiliary/filename_gen.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    :return: The filename.
+    """
+    name = sbom.get("metadata", {}).get("component", {}).get("name", "")
+    name = _sanitize(name)
+    version = sbom.get("metadata", {}).get("component", {}).get("version", "")
+    version = _sanitize(version)
+    timestamp_str: t.Union[str, None] = sbom.get("metadata", {}).get("timestamp")
+
+    if not name and not version and not timestamp_str:
```
[the remark this answers is not in the document]
Comment (mmarseu) [reply cdxev/auxiliary/filename_gen.py:29]: This is the way I intended it. The wording in the documentation _doesn't contain any_ means _contains none_. I could rewrite this to be more clear, if you like.

In a28c847 I added more tests for these cases. As always, the tests document the intention of the code.

**Comment by mmarseu** [comment]
> Just two little comments. But I am wondering why you decided to make `version` not mandatory? Version missing means invalid SBOM.

I'm not considering _version_  to be mandatory because that's only true for Festo-compliant SBOMs.

**Comment by italvi** [comment]
> > Just two little comments. But I am wondering why you decided to make `version` not mandatory? Version missing means invalid SBOM.
> 
> I'm not considering _version_ to be mandatory because that's only true for Festo-compliant SBOMs.

That's not entirely true. Up to cdx 1.3 the `version` was mandatory, too. And according to the [NTIA minimum elements](https://www.ntia.gov/files/ntia/publications/sbom_minimum_elements_report.pdf) it is also mandatory. So I would not see it as a Festo-specific solution.

**Comment by mmarseu** [comment]
> > > Just two little comments. But I am wondering why you decided to make `version` not mandatory? Version missing means invalid SBOM.
> > 
> > 
> > I'm not considering _version_ to be mandatory because that's only true for Festo-compliant SBOMs.
> 
> That's not entirely true. Up to cdx 1.3 the `version` was mandatory, too. And according to the [NTIA minimum elements](https://www.ntia.gov/files/ntia/publications/sbom_minimum_elements_report.pdf) it is also mandatory. So I would not see it as a Festo-specific solution.

So do you **want** the tool to fail when faced with a valid CycloneDX file only because some government agency in some country says you need a version, even if its not mandatory? 😛 

Besides, there are valid use-cases where you would use this tool in an intermediate step with an incomplete SBOM that's still missing some information. I do that in my own pipelines. I'm not using auto-generated file names in those cases but I might as well have.

**Comment by italvi** [comment]
> So do you **want** the tool to fail when faced with a valid CycloneDX file only because some government agency in some country says you need a version, even if its not mandatory? 😛

In that special case, I would say yes, as version should be mandatory. Otherwise it makes different use-cases not feasible. However, do to your comment afterwards:

> Besides, there are valid use-cases where you would use this tool in an intermediate step with an incomplete SBOM that's still missing some information. I do that in my own pipelines. I'm not using auto-generated file names in those cases but I might as well have.

I would say it's fine and we can keep it this way. Would have just reduced the amount of different cases.

File: cdxev/auxiliary/filename_gen.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    Creates a regular expression which can be used to validate the filename of the given SBOM.
+
+    The pattern allows the following variants of the filename:
+
+    * ``bom.json`` is always an allowed name.
+    * ``<name>_[<version>_][<hash>_<timestamp>|<hash>|<timestamp>].cdx.json``, where
+      * ``<name>`` == ``metadata.component.name``, if it exists, otherwise ``unknown``.
+      * ``<version>`` MUST be present, if and only if ``metadata.component.version`` exists.
+      * ``<hash>`` is ``metadata.component.hashes[x].content`` for
```
Comment (italvi) [inline cdxev/auxiliary/filename_gen.py:62]: for what?

Comment (mmarseu) [reply cdxev/auxiliary/filename_gen.py]: That's what happens when you go on lunch break mid-sentence 😆 
Fixed in 25f41ea
